src/ragas/testset/testset_generator.py

```diff
@@ -0,0 +1,309 @@
+        return self._question_transformation(COMPRESS_QUESTION, question=question)
+
+    def _conversational_question(self, question: str):
+        return self._question_transformation(CONVERSATION_QUESTION, question=question)
+
+    def _question_transformation(self, prompt, question: str):
+        human_prompt = prompt.format(question=question)
+        prompt = ChatPromptTemplate.from_messages([human_prompt])
+        results = generate(prompts=[prompt], llm=self.generator_llm)
+        return results.generations[0][0].text.strip()
+
+    def _qc_template(self, prompt, question, context):
+        human_prompt = prompt.format(question=question, context=context)
+        prompt = ChatPromptTemplate.from_messages([human_prompt])
+        results = generate(prompts=[prompt], llm=self.generator_llm)
+        return results.generations[0][0].text.strip()
+
+    def _generate_answer(self, question: str, context: list[str]):
+        return [
+            self._qc_template(ANSWER_FORMULATE, qstn, context[i])
+            for i, qstn in enumerate(question.split("\n"))
+        ]
+
+    def _generate_context(self, question: str, text_chunk: str):
+        return [
+            self._qc_template(CONTEXT_FORMULATE, qstn, text_chunk)
+            for qstn in question.split("\n")
+        ]
+
+    def _remove_index(self, available_indices: list, node_idx: list):
+        for idx in node_idx:
+            available_indices.remove(idx)
+        return available_indices
+
+    def _generate_doc_node_map(self, documenet_nodes: t.List[TextNode]):
+        doc_nodeidx = defaultdict(list)
+        for idx, node in enumerate(documenet_nodes):
+            doc_nodeidx[node.id_].append(idx)
+
+        return doc_nodeidx
+
+    def _get_neighbour_node(self, idx: int, node_indices: list):
+        return [idx - 1, idx] if idx == node_indices[-1] else [idx, idx + 1]
+
+    def _embed_nodes(self, nodes: t.List[TextNode]):
+        embeddings = {}
+        for node in nodes:
+            embeddings[node.id_].update(
+                self.embedding_model.embed_query(node.get_context())
+            )
+
+        return embeddings
+
+    def generate(self, documents: t.List[Document], test_size: int):
+        node_parser = SimpleNodeParser.from_defaults(
+            chunk_size=self.chunk_size, chunk_overlap=0, include_metadata=True
+        )
+        document_nodes = node_parser.get_nodes_from_documents(documents=documents)
+
+        if test_size > len(document_nodes):
+            raise ValueError(
+                """Maximum possible number of samples exceeded, 
+                             reduce test_size or add more documents"""
+            )
+
+        available_indices = np.arange(0, len(document_nodes)).tolist()
+        doc_nodeidx = self._generate_doc_node_map(document_nodes)
+        count = 0
+        Testdata_tuple = namedtuple(
+            "Testdata_tuple", ["question", "context", "answer", "question_type"]
+        )
```

Comment:
we can declare this outside as `TestdataRow` or something so that it can be used as a type